src/coreclr/vm/appdomain.cpp

```diff
@@ -5115,6 +5084,22 @@ HRESULT RuntimeInvokeHostAssemblyResolver(INT_PTR pManagedAssemblyLoadContextToB
                 COMPlusThrowHR(COR_E_INVALIDOPERATION, IDS_HOST_ASSEMBLY_RESOLVER_DYNAMICALLY_EMITTED_ASSEMBLIES_UNSUPPORTED, name);
             }
 
+            // For collectible assemblies, ensure that the parent loader allocator keeps the assembly's loader allocator
+            // alive for all its lifetime.
+            if (pDomainAssembly->IsCollectible())
+            {
+                LoaderAllocator *pResultAssemblyLoaderAllocator = pDomainAssembly->GetLoaderAllocator();
+                LoaderAllocator *pParentLoaderAllocator = pBinder->GetLoaderAllocator();
+                if (pParentLoaderAllocator == NULL)
+                {
+                    // The AssemblyLoadContext for which we are resolving the the Assembly is not collectible.
+                    COMPlusThrow(kNotSupportedException, W("NotSupported_CollectibleBoundNonCollectible")); 
```

Comment:
We should add a test for this (fine with doing it later).

Reply:
Makes sense. I can extend the test I've added previously  to have new test case(s) for that.

Reply:
Actually, I think this is technically a breaking change? I agree it is the right thing, but I'm not sure what the appetite for backporting to 6.0 would be (all about it for 7.0 though).

Reply:
For the 6.0, I'll let the shiproom decide whether to do that or just omit the check. You are right that for some apps, this might accidentally work without this check and now it would always throw an exception.

Reply:
Sounds good to me.

Reply:
Done